File: Trabajo_grupal/WG5/Grupo2_ tarea 5.py
#!/usr/bin/env python
# coding: utf-8

# ### Parte de la Tarea 5 - SPPS File

# In[27]:


pip install pyreadstat

# Tarea 5 - SPSS File
    # Integrantes:
       # Enrique Ríos 
       # Fabio Salas
       # Amalia Castillo
       # Angie  Quispe

# Usted debe trabajar una base data_administrativa que está en la carpeta data. Esta base de datos contiene es una encuesta realizado a miembros del hogar durante años años 2019 y 2020.
# Mostrar las variables que presentan missing values
# Se le pide mostrar las etiquetas de dos variables (var labels) y las etiquetas de los valores en dos variables (value's labels).
# Se le pide detectar personas que fueran entrevistadas en ambos años. Para ello, se pide detectar duplicados a partir del identificador por persona : conglome, vivienda, hogar y codperso.
# Ordene la base de datos a partir de las variables que identifican cada miembro y la variable de año (year). Así podrá observar a cada individuo en ambos años.
# Finalmente crear una base de datos para cada año y guardar en la carpeta data con los siguientes nombres data_2019_(numero de grupo) y data_2020_(numero de grupo).


# In[29]:


import numpy as np
import pandas as pd
import os
import pyreadstat as py

user = os.getlogin()   # Username

# Setear directorio


# ### Se importa los datos a analizar

# In[33]:


os.chdir(f"/Users/enriquerios/Desktop/PUCP 2022.2/R y Python/1ECO35_2022_2/data") # Fijar directoriO


# In[34]:


df,meta=py.read_sav('data_administrativa.sav')


# ### Se revisa la dimensión de la data

# In[35]:


df.shape


# ### Se evalua los datos faltantes

# In[36]:


vacios = pd.DataFrame(df.isnull().sum()).sort_values(0,ascending=False)
vacios.columns = ['vacios']
vacios['vacios%'] = round(vacios['vacios']/df.shape[0], 2)*100
vacios # Datos vacios y vacios % en la tabla


# ###  Etiquetas de las dos variables: Var labels y values labels

# #### Primero : Value labels

# In[37]:


meta.value_labels['labels0'] # Variable geográficaa


# In[38]:


meta.value_labels['labels1'] # variable de habitantes


# #### Segundo: Variable labels

# In[39]:


meta.variable_value_labels['DOMINIO'] #variable del dominio


# In[40]:


meta.variable_value_labels['ESTRATO'] # ahora es variable denominada estrato


# ### Se procede a  detectar y eliminar los duplicados

# In[41]:


m=df.duplicated(df.columns[~df.columns.isin(['conglome','vivienda','hogar','codperso'])],
                        keep='first')


# In[42]:


m_true=m[m==True]


# #### Se encuentra que NO existen duplicados

# In[43]:


m_true


# #### Código para borrar duplicados

# In[44]:


# df = df.drop_duplicates(df.columns[~df.columns.isin(['conglome','vivienda','hogar','codperso'])],
#                        keep='first')
# Se agregó el código para borrar duplicados


# ### Creación de dataframe 2019 y 2020

# In[45]:


df2019=df.copy()
df2020=df.copy()


# In[46]:


df2019.drop(df[df['year']=='2020'].index,inplace=True)
df2020.drop(df[df['year']=='2019'].index,inplace=True)


# In[47]:


df2019.head()


# In[48]:


df2020.head()


# #### Guardamos la nueva data creada

# In[49]:


py.write_sav(df2019,'data_2019_2.sav')
py.write_sav(df2020,'data_2020_2.sav')


# In[ ]:





# In[ ]:

File: Trabajo_grupal/WG5/Grupo 2_ tarea5.R
#-----------------------------------------
# WG5 Tarea 5 - Grupo 2
#-----------------------------------------
# R y Python
#-----------------------------------------
# Integrantes:
# Enrique Ríos 
# Fabio Salas
# Amalia Castillo
# Angie  Quispe

user <-Sys.getenv("USERNAME")
setwd( paste0("C:/Users/",user,"/Documents/GitHub/1ECO35_2022_2/Trabajo_grupal/WG5") )


#--------------
# Comando para borrar todo lo que no sirva antes
rm(list=ls())

#---------------
# Librerías
#---------------
install.packages("pacman")
library(foreign)
library(dplyr)

pacman::p_load(tidyverse, haven, dplyr, janitor , stringr)

# haven: Este comando lee archivos spss (sav)
# string : Este comando trabaja con string

#-----------------------
# Ruta de trabajo
#-----------------------
# Se setea el directorio
script.path <- dirname(rstudioapi::getSourceEditorContext()$path)
setwd("D:\\Users\\Usuario\\Documents\\GitHub\\1ECO35_2022_2\\data")


# Procedemos a importar los datos a analizar

file_path = "D:\\Users\\Usuario\\Documents\\GitHub\\1ECO35_2022_2\\data\\data_administrativa.sav"
datos <- haven::read_sav(file_path , encoding = "UTF-8" )  # read dataset

#Revisamos la dimensión de los datos

head(datos)
names(datos) #Para el nombre de las variables de la lista 


# Var labels y values labels
# Se emplea el comando "control + shift + m " en iOS para que aparezca este símbolo %>% 

datos$ESTRATO %>% attr('labels') # value labels
datos$ESTRATO %>% attr('label') # var label


datos$DOMINIO %>% attr('labels') # value labels
datos$DOMINIO %>% attr('label') # var label

# Primero se procede a desarrollar value labels

print(datos$ESTRATO) #vemos el estrato
datos$ESTRATO %>% attr('labels') # value labels


# Segundo se procede a encontrar las variables labels
print(datos$DOMINIO)
datos$DOMINIO %>% attr('labels') # value labels

# Ahora se procede a detectar y eliminar duplicados
attach(datos) 


datos <- datos %>% filter(ESTRATO == 1 & DOMINIO == 1)

view(datos)

datos_filtrados <- datos %>% group_by(year ,MES , UBIGEO ,CONGLOME , CODPERSO, VIVIENDA, HOGAR) %>% 
  mutate(duplicates = n()) %>% filter(duplicates >1) %>%
  select(year ,MES , UBIGEO ,CONGLOME , CODPERSO, VIVIENDA, HOGAR, duplicates ) 

View(datos_filtrados) # No existen duplicados


# Se crea dataframe para los años 2019 y 2020

# Drop duplicates rows (observaciones). Se borra las copias, No las primeras apariciones

datos_2019 <- datos %>% distinct(year = 2019 , MES , UBIGEO ,CONGLOME , CODPERSO, VIVIENDA, HOGAR) # .keep_all = TRUE muestra todas las variables

datos_2020 <- datos %>% distinct(year = 2020 , MES , UBIGEO ,CONGLOME , CODPERSO, VIVIENDA, HOGAR)

# Finalmente, se guarda la nueva data creada

datos_2019

datos_2020

write_sav(datos_2019, "/Users/enriquerios/Desktop/PUCP 2022.2/R y Python/1ECO35_2022_2/data")
write_sav(datos_2020, "/Users/enriquerios/Desktop/PUCP 2022.2/R y Python/1ECO35_2022_2/data")

